gh-badges/lib/badge-renderers.js

```diff
@@ -0,0 +1,596 @@
+      .replace(/'/g, '&apos;')
+  }
+}
+
+function roundUpToOdd(val) {
+  // Increase chances of pixel grid alignment.
+  return val % 2 === 0 ? val + 1 : val
+}
+
+function preferredWidthOf(str) {
+  return roundUpToOdd((anafanafo(str) / 10) | 0)
+}
+
+function computeWidths({ label, message }) {
+  return {
+    labelWidth: preferredWidthOf(label),
+    messageWidth: preferredWidthOf(message),
+  }
+}
+
+function renderLogo({
+  logo,
+  logoWidth = 14,
+  logoPadding = 0,
+  extraPadding = 0,
+}) {
+  if (!logo) {
+    return {
+      hasLogo: false,
+      totalLogoWidth: 0,
+      renderedLogo: '',
+    }
+  }
+
+  const x = 5 + extraPadding
+  const y = 3 + extraPadding
+  return {
+    hasLogo: true,
+    totalLogoWidth: logoWidth + logoPadding,
+    renderedLogo: `<image x="${x}" y="${y}" width="${logoWidth}" height="14" xlink:href="${escapeXml(
+      logo
+    )}"/>`,
+  }
+}
+
+function renderText({
+  leftMargin,
+  horizPadding = 0,
+  content,
+  verticalMargin = 0,
+  shadow = false,
+}) {
+  if (!content.length) {
+    return { renderedText: '', width: 0 }
+  }
+
+  const textLength = preferredWidthOf(content)
+  const escapedContent = escapeXml(content)
+
+  const shadowMargin = 150 + verticalMargin
+  const textMargin = 140 + verticalMargin
+
+  const outTextLength = 10 * textLength
+  const x = 10 * (leftMargin + 0.5 * textLength + horizPadding)
+
+  let renderedText = ''
+  if (shadow) {
+    renderedText = `<text x="${x}" y="${shadowMargin}" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="${outTextLength}">${escapedContent}</text>`
+  }
+  renderedText += `<text x="${x}" y="${textMargin}" transform="scale(.1)" textLength="${outTextLength}">${escapedContent}</text>`
+
+  return {
+    renderedText,
+    width: textLength,
+  }
+}
+
+function renderLinks({
+  links: [leftLink, rightLink] = [],
+  labelWidth,
+  messageWidth,
+  height,
+}) {
+  leftLink = escapeXml(leftLink)
+  rightLink = escapeXml(rightLink)
+  const hasLeftLink = leftLink && leftLink.length
+  const hasRightLink = rightLink && rightLink.length
+  const leftLinkWidth = hasRightLink ? labelWidth : labelWidth + messageWidth
+
+  function render({ link, width }) {
+    return `<a target="_blank" xlink:href="${link}"><rect width="${width}" height="${height}" fill="rgba(0,0,0,0)"/></a>`
+  }
+
+  return (
+    (hasRightLink
+      ? render({ link: rightLink, width: labelWidth + messageWidth })
+      : '') +
+    (hasLeftLink ? render({ link: leftLink, width: leftLinkWidth }) : '')
+  )
+}
+
+function renderBadge({ links, leftWidth, rightWidth, height }, main) {
+  const width = leftWidth + rightWidth
+  return `
+    <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${width}" height="${height}">
+    ${main}
+    ${renderLinks({ links, leftWidth, rightWidth, height })}
+    </svg>`
+}
+
+function stripXmlWhitespace(xml) {
+  return xml
+    .replace(/>\s+/g, '>')
+    .replace(/<\s+/g, '<')
```

Comment:
I'm not 100% sure what the function is for, but would it be better to strip all whitespace before and after the `<>` chars?
```suggestion
    .replace(/\s*<\s*/g, '<')
```

On another note, would this affect someone who wanted to make a badge that said something like ![label | less < more](https://img.shields.io/badge/label-less%20%3C%20more-blue)?

Reply:
> It's been awhile so i'm not 100% sure what the function is for

It allows us to generate

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <linearGradient id="b" x2="0" y2="100%">
    ...
</svg>
```

(or whatever)..but then serve it as

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><linearGradient id="b" x2="0" y2="100%">...</svg>
```

without using SVGO, or otherwise parsing into an XML object and writing back out to a string.


> But would it be better to strip all whitespace before and after the <> chars?

We're calling `.trim()` at the end which will strip any leading/trailing whitespace. Is that what you mean?



> would this affect someone who wanted to make a badge that said something like ![label | less < more](https://img.shields.io/badge/label-less%20%3C%20more-blue)?

I had thought of this. As long as we've run everything through `escapeXml()`, we'll replace `<` with `&lt;` and so on, so its fine as long as everything is correctly escaped. This would cause problems if we've got an unescaped `<` or `>` though

Reply:
> We're calling `.trim()` at the end which will strip any leading/trailing whitespace. Is that what you mean?

Just for the spacing in between,  like if there was an extra space before the `>` it wouldn't be removed with the current code,
```diff
-<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
```
but thinking about it, It's probably not actually an issue seeing as all that is already hard coded.



> I had thought of this. As long as we've run everything through `escapeXml()`, we'll replace `<` with `&lt;` and so on, so its fine as long as everything is correctly escaped. This would cause problems if we've got an unescaped `<` or `>` though

Good point, forgot about this 👍